File: src/wikibasemigrator/web/components/wikibase_item.js
export default {
  template: `
  <div class="container flex flex-col gap-2 mx-auto">
  <h2 class="flex flex-row text-2xl">
    <div v-html="getItemLink(entity.s_id, s_wikibase)"></div>
    <template v-if="entity.t_id == null">→ CREATE new Item</template>
    <template v-else>→ <div v-html="getItemLink(entity.t_id, t_wikibase)"></div></template>
  </h2>
  <div class="container mx-auto rounded-md bg-slate-300 p-2">
    <div class="grid grid-cols-3 gap-2">
      <div class="flex flex-col rounded-md bg-slate-200">
          <p>Label:</p>
          <div class="flex flex-row gap-2">
              <template v-for="lang in Object.keys(entity.labels)"> <div v-html="getLanguageStrng(entity.labels[lang], lang)"></div></template>
          </div>
      </div>
      <div class="flex flex-col rounded-md bg-slate-200">
          <p>Descriptions:</p>
          <div class="flex flex-row gap-2">
              <template v-for="lang in Object.keys(entity.descriptions)"> <div v-html="getLanguageStrng(entity.descriptions[lang], lang)"></div></template>
          </div>
      </div>
      <div class="flex flex-col rounded-md bg-slate-200">
          <p>Aliases:</p>
          <div class="flex flex-row gap-2">
            <template v-for="lang in Object.keys(entity.aliases)">
                <template v-for="alias in entity.aliases[lang]"> 
                    <div v-html="getLanguageStrng(alias, lang)"></div>
                </template>
            </template>
          </div>
      </div>
    </div>
      
  </div>
  <div v-for="claim in entity.claims">
     <div class="container mx-auto rounded-md bg-slate-300 p-2">
        <div class="flex flex-row gap-4">
           <div v-html="getItemLinkMapping(claim.s_id, claim.t_id)"</div>
           <div class="flex flex-col m-1 gap-2 grow">
              <div v-for="snak in claim.snaks">
                 <div class="container flex flex-col gap-1 p-2 rounded-md bg-slate-200 rounded grow"> 
                    <div v-html="renderSnak(snak.mainsnak)"
                 </div>
                 <div class="ml-10 bg-slate-100 rounded m-1 p-1">
                    <p>Qualifier:</p>
                    <div class="flex flex-col gap-1">
                       <div v-for="qualifier in snak.qualifiers" class="grid grid-cols-2  gap-5 px-1  rounded ring-1 m-1 ">
                          <div v-html="getItemLinkMapping(qualifier.s_id, qualifier.t_id)" class="flex flex-row"></div>
                          <div class="flex flex-col gap-1 grow">
                             <div v-for="snak in qualifier.snaks">
                                <div v-html="renderSnak(snak)"></div>
                             </div>
                          </div>
                       </div>
                    </div>
                 </div>
                 <div class="flex flex-col gap-2 ml-10 m-1 p-1  rounded bg-slate-100">
                    <div class="">References:</div>
                    <div v-for="reference_block in snak.references" >
                       <div class="m-1100 rounded ring-1 p-1">
                          <div v-for="reference in reference_block.reference" class="grid grid-cols-2  gap-5 px-1 bg-slate-50 m-1 rounded ring-1 ">
                             <div v-html="getItemLinkMapping(reference.s_id, reference.t_id)" class="flex flex-row"></div>
                             <div class="flex flex-col gap-2 grow">
                                <div v-for="snak in reference.snaks">
                                   <div v-html="renderSnak(snak)"></div>
                                </div>
                             </div>
                          </div>
                       </div>
                    </div>
                 </div>
              </div>
           </div>
        </div>
     </div>
  </div>
  </div>
  </div>
</div>
  `,
  data() {
    return {
      value: 0,
      css_classes_link:"underline text-blue-600 hover:text-blue-800 visited:text-purple-600"
    };
  },
  methods: {
    getSourceLabel(id) {
      return this.source_labels[id] || "no label"
    },
    getTargetLabel(id) {
      return this.target_labels[id] || "no label"
    },
    getLanguageStrng(str, lang){
      return `"${str}"@${lang}`
    },
    renderSnak(snak){
      if (snak.snaktype !== "value"){
        return `<div> ${snak.value} ${this.getItemLink("(build-in type)", "https://www.mediawiki.org/wiki/Wikibase/DataModel#Snaks")}</div`
      }
      switch (snak.type) {
        case "wikibase-item":
          return this.getItemLinkMapping(snak.value, snak.target_value)
        case "monolingualtext":
          return this.getLanguageStrng(snak.value, snak.language)
        default:
          return `<div> "${snak.value}"</div`
      }
    },
    getItemLink(item, location){
      return `<a href="${location}${item}" class="${this.css_classes_link}">${item}</a>`
    },
    getItemLinkMapping(s_id, t_id){
      if(t_id == null){
        return `
            <a href="${this.s_wikibase}${s_id}" class="${this.css_classes_link}">${s_id} (${ this.getSourceLabel(s_id) })</a>
            →
            <div class="text-red">MISSING</div>
        `
      }else{
          return `
            <a href="${this.s_wikibase}${s_id}" class="${this.css_classes_link}">${s_id} (${this.getSourceLabel(s_id)})</a>
            →
            <a href="${this.t_wikibase}${t_id}" class="${this.css_classes_link}">${t_id} (${this.getTargetLabel(t_id)})</a>
        `
        }
    },
  },
  props: {
    s_wikibase: URL,
    t_wikibase: URL,
    source_labels: Object,
    target_labels: Object,
    entity: Object,
    css_classes_link: String
  },
};

File: src/wikibasemigrator/web/components/wikibase_item.py
from nicegui.element import Element
from pydantic import HttpUrl
from wikibaseintegrator.models import Reference, Snak
from wikibaseintegrator.wbi_enums import WikibaseSnakType

from wikibasemigrator.model.translations import EntityTranslationResult


class TranslatedWikibaseItemWidget(Element, component="wikibase_item.js"):
    def __init__(
        self,
        source_url: str | HttpUrl,
        target_url: str | HttpUrl,
        source_labels: dict[str, str] | None,
        target_labels: dict[str, str] | None,
        translation_result: EntityTranslationResult | None = None,
        entity: dict | None = None,
    ) -> None:
        super().__init__()
        if translation_result is not None:
            entity = self._convert_to_translation_record(translation_result)
        self._props["s_wikibase"] = str(source_url)
        self._props["t_wikibase"] = str(target_url)
        self._props["source_labels"] = source_labels
        self._props["target_labels"] = target_labels
        self._props["entity"] = entity

    def _convert_to_translation_record(self, translation_result: EntityTranslationResult) -> dict:
        """
        Conver the translation result to a translation record which can be rendered by the vue template
        :param translation_result:
        :return:
        """
        item = translation_result.original_entity
        mappings = translation_result.entity_mapping
        claims: dict[str, list] = {}
        for claim in item.claims:
            mainsnak = self._convert_snak(claim.mainsnak, mappings)
            qulifier_recors = []
            reference_records = []
            for property_number, snaks in claim.qualifiers.qualifiers.items():
                qualifier_record = {
                    "s_id": property_number,
                    "t_id": mappings.get(property_number, None),
                    "snaks": [self._convert_snak(snak, mappings) for snak in snaks],
                }
                qulifier_recors.append(qualifier_record)
            reference_block: Reference
            for reference_block in claim.references:
                block_records = []
                for property_number, snaks in reference_block.snaks.snaks.items():
                    reference_record = {
                        "s_id": property_number,
                        "t_id": mappings.get(property_number, None),
                        "snaks": [self._convert_snak(snak, mappings) for snak in snaks],
                    }
                    block_records.append(reference_record)
                reference_records.append({"reference": block_records})
            claim_record = {"mainsnak": mainsnak, "qualifiers": qulifier_recors, "references": reference_records}
            if claim.mainsnak.property_number not in claims:
                claims[claim.mainsnak.property_number] = []
            claims[claim.mainsnak.property_number].append(claim_record)
        claim_records: list[dict[str, str | list | None]] = [
            {"s_id": s_id, "t_id": mappings.get(s_id, None), "snaks": snaks} for s_id, snaks in claims.items()
        ]
        # Sort claims: properties with mapping first than sort by property number
        claim_records.sort(
            key=lambda x: int(f"1{x.get('s_id', '').removeprefix('P')}")  # type: ignore
            if x.get("t_id") is not None
            else int(f"9{x.get('s_id', '').removeprefix('P')}")  # type: ignore
        )
        entity_record = {
            "s_id": item.id,
            "t_id": mappings.get(item.id, None),
            "labels": {label.language: label.value for label in translation_result.entity.labels},
            "descriptions": {desc.language: desc.value for desc in translation_result.entity.descriptions},
            "aliases": {
                language: [alias.value for alias in aliases]
                for language, aliases in translation_result.entity.aliases.aliases.items()
            },
            "claims": claim_records,
        }
        return entity_record

    def _convert_snak(self, snak: Snak, mappings: dict) -> dict:
        record = {"type": snak.datatype, "snaktype": snak.snaktype.value}
        # ToDo: refactor all to .get()
        if snak.snaktype is WikibaseSnakType.UNKNOWN_VALUE:
            record["value"] = WikibaseSnakType.UNKNOWN_VALUE.value
        elif snak.snaktype is WikibaseSnakType.NO_VALUE:
            record["value"] = WikibaseSnakType.NO_VALUE.value
        else:
            match snak.datatype:
                case (
                    "string"
                    | "external-id"
                    | "commonsMedia"
                    | "entity-schema"
                    | "url"
                    | "property"
                    | "geo-shape"
                    | "tabular-data"
                ):
                    record["value"] = snak.datavalue["value"]
                case "wikibase-item":
                    record["value"] = snak.datavalue["value"]["id"]
                    record["target_value"] = mappings.get(snak.datavalue["value"]["id"], None)
                case "time":
                    record["value"] = snak.datavalue["value"]["time"]
                case "quantity":
                    record["value"] = f"""{snak.datavalue}"""
                case "monolingualtext":
                    record["value"] = snak.datavalue["value"].get("text", None)
                    record["language"] = snak.datavalue["value"].get("language", None)
                case "globecoordinate":
                    record["value"] = str(snak.datavalue["value"])
        return record
